src/types.py

```diff
@@ -69,16 +69,29 @@ class Type:
 
     _data: TypeData = attr.ib()
 
-    def unify(self, other: "Type") -> bool:
+    def unify(self, other: "Type", *, seen: Optional[Set["TypeData"]] = None) -> bool:
         """
         Try to set this type equal to another. Returns true on success.
         Once set equal, the types will always be equal (we use a union-find
         structure to ensure this).
+        The seen argument is used during recursion, to track which TypeData
+        objects have been encountered so far.
         """
+
         x = self.data()
         y = other.data()
         if x is y:
             return True
+
+        # If we hit a type that we have already seen, fail.
+        # TODO: Is there a looser check that would allow more types to unify?
```

Comment:
I think in principle it should be possible to recursively unify arbitrary types by moving recursive calls to after the union-find merge step, and then adding the revert machinery we talked about. Maybe worth doing just for fun, though I doubt there's any practical value to it.